### PR TITLE
PUBDEV-4507: XGBoost integration to AutoML

### PR DESCRIPTION
**Pull request by seb-h2o**
https://0xdata.atlassian.net/browse/PUBDEV-4507
- tuned 3 default XGBoost models for small to larger datasets.
- added them before existing default GBMs.
- added XGBoost hyperparam search before GBM one.
- based on XGBoost, added experimental support for emulated LightGBM (currently disabled as I expected faster perf according to https://lightgbm.readthedocs.io/en/latest/Experiments.html#result)
- orchestration of various hyperparam searches to ensure that one is not monopolising resources (remaining models + time), and ensure we have a balance between XGB, GBM and DL.

### CODE REVIEW

File: h2o-automl/src/main/java/ai/h2o/automl/AutoML.java [hunk 1]
```diff
@@ -41,27 +43,113 @@
...
   // NOTE: make sure that this is in sync with the exclude option in AutoMLBuildSpecV99
   public enum Algo {
     GLM,
     DRF,
     GBM,
     DeepLearning,
     StackedEnsemble,
-//    XGBoost,
+    XGBoost,
+    LightGBM { @Override String urlName() { return XGBoost.urlName(); } }
```
Comment (ledell) [inline h2o-automl/src/main/java/ai/h2o/automl/AutoML.java]: Let's fold the pseudo-LightGBM models into "XGBoost" for now and remove the `LightGBM` option.

**Comment by seb-h2o** [comment]
@ledell do you see anything else that could be done on this PR?
want me to add some results on various datasets to the ticket? already added MiniBoone results to https://0xdata.atlassian.net/browse/PUBDEV-4507.
Now I can try Albert with local fix... (still binomial, but >2x larger).

Otherwise, I can't see any necessary XGBoost-specific work for this integration, do you?

**Comment by seb-h2o** [comment]
> Hey @seb-h2o We should benchmark this against the existing version. Have you had a chance to run any benchmarks yet?

The only benchmarks I'm aware of for AutoML are run against rel-* version, not against master: talking about those http://benchmark:8050/.
In H2O-3, I can't find any AutoML benchmarks yet... Should we add one to the Jenkins pipeline using same datasets as for XGB or GBM?

Also, I'd be interested in discussing what we want to measure exactly in those benchmarks:
- time perf doesn't look very relevant given that AutoML is timebound and that DL grids can easily run for a long time without always providing good accuracy.
- should we measure accuracies based both on validation dataset + test dataset to measure overfitting?
- should we measure diversity of algorithms produced: I don't really like the approach of only taking the leader and check its performance. I think we should measure the performance distribution of all generated models, this would give a better view of AutoML as whole than just its "best" model.

If some benchmark work has been started for AutoML, thanks for pointing me to it, otherwise I'll try to integrate sth to Jenkins pipeline.

File: h2o-automl/src/main/java/ai/h2o/automl/AutoML.java [hunk 2]
```diff
@@ -42,27 +44,127 @@
...
+   */
+  @Deprecated
+  public interface algo {
+    @Deprecated public static final Algo GLM = Algo.GLM;
+    @Deprecated public static final Algo DRF = Algo.DRF;
+    @Deprecated public static final Algo GBM = Algo.GBM;
+    @Deprecated public static final Algo DeepLearning = Algo.DeepLearning;
+    @Deprecated public static final Algo StackedEnsemble = Algo.StackedEnsemble;
+    @Deprecated public static final Algo XGBoost = Algo.XGBoost;
+  }
```
Comment (seb-h2o) [inline h2o-automl/src/main/java/ai/h2o/automl/AutoML.java]: @michalkurka : I had many conflicts and compilation issues (`algo` being used as a class and as a variable in many places then) when rebasing the renaming changes on this branch.
I suggest the above deprecation to be used for external legacy use only, besides deprecation will inform the user so that we will be able to remove it safely at some point.
WDYT?

File: h2o-automl/src/main/java/ai/h2o/automl/AutoML.java [hunk 3]
```diff
@@ -42,27 +44,127 @@
...
     ;
 
     String urlName() {
       return this.name().toLowerCase();
     }
   }
 
+  /**
+   * for external use only.
+   * Keeping this mapping for customers consuming the Java API directly
```
Comment (michalkurka) [inline h2o-automl/src/main/java/ai/h2o/automl/AutoML.java]: Please use the term "users" instead of "customers". Can you also include a note about when will this compatibility interface be removed? Eg.: deprecated, will be removed in H2O 3.24.0.1, please use ...

File: h2o-automl/build.gradle [hunk 4]
```diff
@@ -4,6 +4,7 @@ dependencies {
   compile project(":h2o-genmodel")
   compile project(":h2o-core")
   compile project(":h2o-algos")
+  compile project(":h2o-ext-xgboost")
```
Comment (michalkurka) [inline h2o-automl/build.gradle]: I don't think this is an issue for H2O deployments but I wanted to point out that xgboost is designed as an optional extension. It is bundled in H2O, h2odriver and SW but users might choose not to include xgboost in their custom H2O builds.
It would be better if xgboost was loaded dynamically based on availability but I do understand that would be a bigger change and this is fine for now.

Comment (seb-h2o) [reply h2o-automl/build.gradle]: I just need XGBoostParameters class when training XGBoost model and no class from XGBoost extension is loaded before that, indeed there XGBoost extension is first detected using ExtensionManager and no XGBoost-related code should be executed if the extension is missing.
Having XGBoostParameters in the imports should be irrelevant for the classloader, so I think we're fine already. Need to double-check though.

**Comment by seb-h2o** [comment]
> Why Serializable? is that something for SW? It is a bit unusual to see Serializable in H2O ;)

it is, right? :) it's needed by the Icer/Weaver as the AutoMLBuildSpec declares exclude_algos as an Automl.algo for backwards compatibility. Enums being Serializable by default, I had to add it for this old-style enum interface.

Also going to remove XGBoost from this enum, this will encourage users to migrate to new enum when they need to modify their code because of XGBoost.

**Comment by seb-h2o** [comment]
@michalk, @erin
this is ready for merge to master, latest changes FYI:
- AutoML-specific de/activation logic for XGBoost + doc.
- early parameter validation for client-mode in ModelBuilder to handle exception smoothly.

**Comment by seb-h2o** [comment]
@michalkurka regarding the _ coding style, I agree that it's something that could be refactored quickly later to align with core.
